Vectorize elu and its backward function on CPU

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32986 Vectorize elu and its backward function on CPU**

Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.292799739996553
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.828570917001343
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1359513780043926
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7030876770004397
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.568238995998399
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8908141480060294
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.8652471189998323
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.13068484600808
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
2.1265591429983033
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.6708065870043356
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.1806934149935842
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.77735430400935
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.494567882007686
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007220732004498
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.615133151994087
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.105554559995653
```

Differential Revision: [D19794595](https://our.internmc.facebook.com/intern/diff/D19794595)

xuhdev committed:
Vectorize elu and its backward function

Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.455715415999293
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.9717224089981755
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.05698063201271
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.554642122995574
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.319330166996224
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007115418993635
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.562178406995372
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.0530152919964166
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
1.9431513289891882
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5050278619892197
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0582968829985475
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6583983750024345
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.37903741998889
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.882683959993301
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.59183525100525
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.082764150996809
```

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 676ef01c:


**None of the build failures appear to be your fault**.

* **2/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=676ef01cd7910f0407fda4aaedd8c7210f9240f0).

### :snowflake: 2 failures recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4414010) pytorch_linux_xenial_py3_6_gcc5_4_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95196941) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95196941)) :snowflake:
<details>
<summary>
<code>Feb 04 21:20:39 AssertionError: tensor(1.9455) not less than or equal to 0.0001 :</code>
</summary>

```
Feb 04 21:20:39   File "test_nn.py", line 8485, in <lambda> 
Feb 04 21:20:39     add(test_name, lambda self, test=test: test(self)) 
Feb 04 21:20:39   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_nn.py", line 3653, in __call__ 
Feb 04 21:20:39     self.test_noncontig(test_case, module, input) 
Feb 04 21:20:39   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_nn.py", line 3715, in test_noncontig 
Feb 04 21:20:39     test_case.assertEqual(grad, d_input, 1e-4) 
Feb 04 21:20:39   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 862, in assertEqual 
Feb 04 21:20:39     assertTensorsEqual(x, y) 
Feb 04 21:20:39   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 832, in assertTensorsEqual 
Feb 04 21:20:39     self.assertLessEqual(max_err, prec, message) 
Feb 04 21:20:39 AssertionError: tensor(1.9455) not less than or equal to 0.0001 :  
Feb 04 21:20:39  
Feb 04 21:20:40 ---------------------------------------------------------------------- 
Feb 04 21:20:40 Ran 1532 tests in 521.534s 
Feb 04 21:20:40  
Feb 04 21:20:40 FAILED (failures=3, errors=1, skipped=706, expected failures=2) 
Feb 04 21:20:40  
Feb 04 21:20:40 Generating XML reports... 
Feb 04 21:20:40 Traceback (most recent call last): 
Feb 04 21:20:40   File "test/run_test.py", line 457, in <module> 
Feb 04 21:20:40     main() 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4414082) pytorch_linux_xenial_py2_7_9_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=95197222) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=95197222)) :snowflake:
<details>
<summary>
<code>Feb 04 21:21:31 AssertionError: tensor(1.9455) not less than or equal to 0.0001 :</code>
</summary>

```
Feb 04 21:21:31   File "test_nn.py", line 8485, in <lambda> 
Feb 04 21:21:31     add(test_name, lambda self, test=test: test(self)) 
Feb 04 21:21:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_nn.py", line 3653, in __call__ 
Feb 04 21:21:31     self.test_noncontig(test_case, module, input) 
Feb 04 21:21:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_nn.py", line 3715, in test_noncontig 
Feb 04 21:21:31     test_case.assertEqual(grad, d_input, 1e-4) 
Feb 04 21:21:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_utils.py", line 862, in assertEqual 
Feb 04 21:21:31     assertTensorsEqual(x, y) 
Feb 04 21:21:31   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/testing/_internal/common_utils.py", line 832, in assertTensorsEqual 
Feb 04 21:21:31     self.assertLessEqual(max_err, prec, message) 
Feb 04 21:21:31 AssertionError: tensor(1.9455) not less than or equal to 0.0001 :  
Feb 04 21:21:31  
Feb 04 21:21:31 ---------------------------------------------------------------------- 
Feb 04 21:21:31 Ran 1532 tests in 544.147s 
Feb 04 21:21:31  
Feb 04 21:21:31 FAILED (failures=3, errors=1, skipped=724, expected failures=3) 
Feb 04 21:21:31  
Feb 04 21:21:31 Generating XML reports... 
Feb 04 21:21:31 Traceback (most recent call last): 
Feb 04 21:21:31   File "test/run_test.py", line 457, in <module> 
Feb 04 21:21:31     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.455715415999293
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.9717224089981755
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.05698063201271
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.554642122995574
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.319330166996224
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007115418993635
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.562178406995372
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.0530152919964166
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
1.9431513289891882
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5050278619892197
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0582968829985475
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6583983750024345
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.37903741998889
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.882683959993301
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.59183525100525
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.082764150996809
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.455715415999293
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.9717224089981755
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.05698063201271
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.554642122995574
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.319330166996224
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007115418993635
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.562178406995372
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.0530152919964166
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
1.9431513289891882
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5050278619892197
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0582968829985475
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6583983750024345
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.37903741998889
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.882683959993301
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.59183525100525
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.082764150996809
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.324265361996368
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.821558489013114
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1433291809953516
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.703330095988349
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.404582636998384
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.9076431900030002
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.610289914999157
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.1557736020040466
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
1.95096738099528
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5198447719885735
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0382273199938936
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6704767499904847
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.496162325012847
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.9348144579998916
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.6738108650024515
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.0817625069903443
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.334304119998706
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.832098839004175
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.15845200100739
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7132766059949063
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.399590691988124
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.0036709580017487
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.5709401329950197
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.0690161569946213
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
2.0004763710021507
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5047218489926308
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0963611560000572
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6705579239933286
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.495936627004994
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8804640369926346
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.6975312110007508
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.1641451030009193
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.292799739996553
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.828570917001343
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1359513780043926
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7030876770004397
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.568238995998399
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8908141480060294
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.8652471189998323
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.13068484600808
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
1.979430178005714
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.5234245649917284
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.0334660360094858
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.6610885229893029
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.5398021370056085
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.0257887030020356
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.627759244001936
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.1315538480121177
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.292799739996553
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.828570917001343
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1359513780043926
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7030876770004397
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.568238995998399
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8908141480060294
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.8652471189998323
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.13068484600808
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
2.1265591429983033
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.6708065870043356
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.1806934149935842
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.77735430400935
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.494567882007686
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007220732004498
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.615133151994087
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.105554559995653
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.292799739996553
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.828570917001343
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1359513780043926
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7030876770004397
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.568238995998399
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8908141480060294
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.8652471189998323
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.13068484600808
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
2.1265591429983033
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.6708065870043356
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.1806934149935842
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.77735430400935
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.494567882007686
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007220732004498
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.615133151994087
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.105554559995653
```

[ghstack-poisoned]

xuhdev committed:
Update on "Vectorize elu and its backward function on CPU"


Benchmark: (Debian 10, Release build, gcc 8.3, no turbo, Intel(R) Xeon(R) E-2136 CPU @ 3.30GHz)

```python
import timeit
for op in ('ELU',):
    print('Forward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(10_000, 100000),
                    (100_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('m(a)', setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, dtype={dtype})', number=t))
    print('Backward')
    for dtype in ('torch.double', 'torch.float'):
        for n, t in [(20_000, 100000),
                    (200_000, 10000)]:
            print(f'torch.nn.{op}()(a), numel() == {n} for {t} times, dtype={dtype}')
            print(timeit.timeit('y.backward(retain_graph=True)',
                                setup=f'import torch; m = torch.nn.{op}(); a = torch.linspace(-1, 1, {n}, requires_grad=True, dtype={dtype}); x = m(a); y = x.sum()',
                                number=t))
```

Before:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
5.292799739996553
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
4.828570917001343
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
3.1359513780043926
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
2.7030876770004397
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.568238995998399
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
1.8908141480060294
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.8652471189998323
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.13068484600808
```

After:

```
Forward
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.double
2.1265591429983033
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.double
1.6708065870043356
torch.nn.ELU()(a), numel() == 10000 for 100000 times, dtype=torch.float
1.1806934149935842
torch.nn.ELU()(a), numel() == 100000 for 10000 times, dtype=torch.float
0.77735430400935
Backward
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.double
4.494567882007686
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.double
2.007220732004498
torch.nn.ELU()(a), numel() == 20000 for 100000 times, dtype=torch.float
3.615133151994087
torch.nn.ELU()(a), numel() == 200000 for 10000 times, dtype=torch.float
1.105554559995653
```

Differential Revision: [D19794595](https://our.internmc.facebook.com/intern/diff/D19794595)

[ghstack-poisoned]

xuhdev commented:
@VitalyFedyunin Mind reviewing this one? :)
